cpp/src/join/detail/intersection.cuh

```diff
@@ -0,0 +1,144 @@
+
+template <typename InputIterator, typename OutputIterator>
+inline cudf::size_type copy_leaf_intersections(InputIterator input_begin,
+                                               InputIterator input_end,
+                                               OutputIterator output_begin,
+                                               cudaStream_t stream)
+{
+  return thrust::distance(
+    output_begin,
+    thrust::copy_if(rmm::exec_policy(stream)->on(stream),
+                    input_begin,
+                    input_end,
+                    output_begin,
+                    [] __device__(auto const &t) { return thrust::get<0>(t) == leaf_indicator; }));
+}
+
+template <typename InputIterator, typename OutputIterator>
+inline cudf::size_type remove_non_quad_intersections(InputIterator input_begin,
+                                                     InputIterator input_end,
+                                                     OutputIterator output_begin,
+                                                     cudaStream_t stream)
+{
+  return thrust::distance(output_begin,
+                          thrust::remove_if(rmm::exec_policy(stream)->on(stream),
+                                            input_begin,
+                                            input_end,
+                                            output_begin,
+                                            [] __device__(auto const &t) {
+                                              return thrust::get<0>(t) != quad_indicator;
+                                            }));
+}
+
+template <typename T,
+          typename NodeIndicesIterator,
+          typename PolyIndicesIterator,
+          typename NodePairsIterator,
+          typename LeafPairsIterator>
+inline std::pair<cudf::size_type, cudf::size_type> find_intersections(
+  cudf::table_view const &quadtree,
+  cudf::table_view const &poly_bbox,
+  NodeIndicesIterator node_indices,
+  PolyIndicesIterator poly_indices,
+  NodePairsIterator node_pairs,
+  LeafPairsIterator leaf_pairs,
+  cudf::size_type num_pairs,
+  T x_min,
+  T y_min,
+  T scale,
+  cudf::size_type max_depth,
+  cudaStream_t stream)
+{
+  auto nodes_and_polys = make_zip_iterator(
+    node_indices,
+    thrust::make_permutation_iterator(quadtree.column(0).begin<uint32_t>(), node_indices),  // keys
+    thrust::make_permutation_iterator(quadtree.column(1).begin<uint8_t>(), node_indices),  // levels
+    thrust::make_permutation_iterator(quadtree.column(2).begin<bool>(), node_indices),  // is_quad
+    poly_indices,
+    thrust::make_permutation_iterator(poly_bbox.column(0).begin<T>(), poly_indices),   // poly_x_min
+    thrust::make_permutation_iterator(poly_bbox.column(1).begin<T>(), poly_indices),   // poly_y_min
+    thrust::make_permutation_iterator(poly_bbox.column(2).begin<T>(), poly_indices),   // poly_x_max
+    thrust::make_permutation_iterator(poly_bbox.column(3).begin<T>(), poly_indices));  // poly_y_max
```

Comment:
I'm curious whether using `thrust::make_permutation_iterator()` multiple times with the same indices leads to multiple reads from global memory, or if that's cached. If so, I can refactor this to be one `permutation_iterator()` per set of indices like this:
```c++
  auto nodes_and_polys = make_zip_iterator(
    node_indices,
    thrust::make_permutation_iterator(make_zip_iterator(
      quadtree.column(0).begin<uint32_t>(), // keys
      quadtree.column(1).begin<uint8_t>(),  // levels
      quadtree.column(2).begin<bool>()      // is_quad
    ), node_indices),
    poly_indices,
    thrust::make_permutation_iterator(make_zip_iterator(
      poly_bbox.column(0).begin<T>(),   // poly_x_min
      poly_bbox.column(1).begin<T>(),   // poly_y_min
      poly_bbox.column(2).begin<T>(),   // poly_x_max
      poly_bbox.column(3).begin<T>()    // poly_y_max
    ), poly_indices)
  );
```

Reply:
I simplified the logic by doing the lookups inside the `transform` lambda directly.